rust-runtime/aws-smithy-types/src/auth.rs

```diff
@@ -0,0 +1,270 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+//! Smithy HTTP Auth Types
+
+use std::cmp::PartialEq;
+use std::fmt::Debug;
+
+#[derive(Debug, PartialEq)]
+enum AuthErrorKind {
+    InvalidLocation,
+    SchemeNotAllowed,
+}
+
+/// Error for Smithy authentication
+#[derive(Debug)]
+pub struct AuthError {
+    kind: AuthErrorKind,
+}
+
+impl std::fmt::Display for AuthError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        use AuthErrorKind::*;
+        match &self.kind {
+            InvalidLocation => write!(f, "invalid location: expected `header` or `query`"),
+            SchemeNotAllowed => write!(
+                f,
+                "scheme only allowed when it is set into the `Authorization header`"
+            ),
+        }
+    }
+}
+
+impl From<AuthErrorKind> for AuthError {
+    fn from(kind: AuthErrorKind) -> Self {
+        Self { kind }
+    }
+}
+
+/// Authentication configuration to connect to a Smithy Service
+#[derive(Clone, Debug, PartialEq)]
+pub struct AuthApiKey {
+    api_key: String,
```

Comment:
We should use `Zeroizing` on this so that it is wiped from memory when no longer in use (see `Credentials` for example: https://github.com/awslabs/smithy-rs/blob/76ee00eb1737964ce240ae817f573980920f0607/aws/rust-runtime/aws-credential-types/src/credentials_impl.rs#L25).

We also need to implement `Debug` manually so that the API key value can be redacted. Example: https://github.com/awslabs/smithy-rs/blob/76ee00eb1737964ce240ae817f573980920f0607/aws/rust-runtime/aws-credential-types/src/credentials_impl.rs#L47